dartagnan/src/main/java/com/dat3m/dartagnan/witness/graphviz/ExecutionGraphVisualizer.java

```diff
@@ -232,25 +323,34 @@ private String eventToNode(EventData e) {
         return "\"" + nodeString + "\"";
     }
 
-    private void appendEdge(EventData a, EventData b, String... options) {
+    private void appendEdge(EventModel a, EventModel b, String... options) {
         graphviz.addEdge(eventToNode(a), eventToNode(b), options);
     }
 
-    public static File generateGraphvizFile(ExecutionModel model, int iterationCount,
-                                            BiPredicate<EventData, EventData> rfFilter, BiPredicate<EventData, EventData> frFilter,
-                                            BiPredicate<EventData, EventData> coFilter, String directoryName, String fileNameBase,
+    private void appendNode(EventModel e, String... attributes) {
+        graphviz.addNode(eventToNode(e), attributes);
+    }
+
+    public static File generateGraphvizFile(ExecutionModelNext model,
+                                            EncodingContext context,
+                                            Model smtModel,
```

Comment:
Appearance of `Model` can be avoid by changing the method for showing `CO`. Another usage for `EncodingContext` is to get the `Configuration` object to be injected the option `--witness.show`. Maybe it's better to pass the `Configuration` here directly?

Reply:
Yes, pass configuration directly.

Reply:
I think `Model` and `EncodingContext` must be here, at least in this method. Because `ExecutionModelManager` can only build the `ExecutionModelNext` by `buildExecutionModel(context, model)` AFTER the configuration being injected.

Reply:
You are basically saying that the visualizer constructs its own `ExecutionModelNext` based on its configuration, no? If so, why do you pass an `ExecutionModelNext` in the first place?
I mean, either the visualizer gets an execution model as input or it constructs its own one, but it shouldn't do both.

IMO, the easiest and cleanest way would be to simply have `ExecutionModelNext` contain a `RelationModel` for all relations of the memory model so that the visualizer just has to pick what it wants to show.
Alternatively, you have to already respect the `witness.show` configuration in the solver methods, so they know how detailed their produced witnesses needs to be and thus can ensure the relevant relations get computed.

Reply:
`RelationModels` can be constructed on demand for derived relations. For base relations, there is no chance really.

Reply:
>For base relations, there is no chance really.

What do you mean by this?

Reply:
If the `ExecutionModel` does not capture the base relations, then you generally have no chance(*) to obtain them apart from looking into the SMT model again. But if you need to do this, then `ExecutionModel` failed its purpose.

(*) Technically, some "static" relations like `po` and `loc` can still be computed afterwards.

Reply:
The current design is: all components of an ExecutionModelNext are extracted only once. That being said, once an ExecutionModelNext is constructed by the manager with the build() method, it becomes “immutable”, serving as a fixed “snapshot” of the execution generated as your demand (the demand for relations). By default (inside RefinementSolver), PO, CO, RF are extracted, otherwise (inside visualizer) only the relations to be shown are extracted and added to ExecutionModelNext. I think this logic works fine so far.

Reply:
If you only extract a small set of relations on construction that is insufficient to describe the execution, then it means every user has to construct the kind of instance he needs. But that in turn means that you need to pass around the SMT model, which kinda defeats the purpose of `ExecutionModel`.
The point is to have an SMT-agnostic representation of what an execution is. No user (except the `RefinementSolver`) should really care at all if this execution comes from an SMT model or if it was generated in some other way (say by brute-force enumeration, by parsing from a witness file, by an oracle, etc.).

If you allow for partial information in `ExecutionModel` (i.e., extract only a subset of the base relations), then whoever generates this model (i.e., the solvers) is responsible to generate enough information for the user, e.g. the visualizer, to do its job. It should not be responsibility of the user to come up with its own executions.
The easiest way to achieve this is to simply generate complete information about the execution (this would satisfy every possible user).

That being said, it would also be fine for the user to generate "enriched views" of the execution, say, by adding derived relations to it via decorators (or rather asking the `ExecutionModelManager` to do so) like this:
```
ExecutionModel model = ... // only contains base relations
ExecutionModel enrichedModel = ExecutionModelManager.enrichWith(model, <wmm>, <derivedRelations>)
// enrichedModel is a view of the original model with derived relations added
```
I don't require you to implement this decorator features, it should just illustrate that the original model should have enough information to generate this information without(!) referring back to the SMT model.

Reply:
Edited. The `Manager` now extracts all `Relations` and constructs corresponding `RelationModels` so that all users get the same and complete `ExecutionModel`, and passing around `EncodingContext` and `Model` is avoided.